src/Datadog.Trace.ClrProfiler.Native/clr_helpers.cpp

```diff
@@ -682,4 +683,212 @@ bool TryParseSignatureTypes(const ComPtr<IMetaDataImport2>& metadata_import,
+          return false;
+      }
+
+      if (FAILED(hr)) {
+        Warn("[trace::ReturnTypeIsValueTypeOrGeneric] element_type=", ret_type, ": failed to get parent token or signature");
+        return false;
+      }
+
+      // Determine the index of GenArgCount in the signature
+      size_t parent_token_index;
+      if (token_type == mdtMemberRef || token_type == mdtMethodDef) {
+        // TypeSpec Format: GENERICINST (CLASS | VALUETYPE) TypeDefOrRefEncoded GenArgCount Type Type*
+        // Skip over TypeDefOrRefEncoded by parsing the signature at index 2
+        parent_token_index = 2;
+        mdToken dummy_token;
+        ULONG token_length = CorSigUncompressToken(
+            &spec_signature[parent_token_index], &dummy_token);
+        parent_token_index += token_length;
+      } else if (token_type == mdtMethodSpec) {
+        // MethodSpec Format: GENRICINST GenArgCount Type Type*
+        parent_token_index = 1;
+      } else {
+        Warn("[trace::ReturnTypeIsValueTypeOrGeneric] element_type=", ret_type, ": token_type (", token_type , ") not recognized");
+        return false;
+      }
+
+      // Read the value of GenArgCount in the signature
+      ULONG num_generic_arguments;
+      parent_token_index += CorSigUncompressData(
+          &spec_signature[parent_token_index], &num_generic_arguments);
+
+      // Get a pointer to first type after GenArgCount that we can increment to read the signature
+      PCCOR_SIGNATURE p_current_byte = spec_signature + parent_token_index;
+
+      // Iterate to specified generic type argument index and return the appropriate class token or TypeSpec
+      for (size_t i = 0; i < num_generic_arguments; i++) {
+        if (i != generic_type_index) {
+          if (!ParseType(&p_current_byte)) {
+            Warn(
+                "[trace::ReturnTypeIsValueTypeOrGeneric] element_type=", ret_type, ": Unable to parse "
+                "generic type argument ", i,
+                "from signature of parent_token:", parent_token);
+            return false;
+          }
+        } else if (*p_current_byte == ELEMENT_TYPE_MVAR ||
+                    *p_current_byte == ELEMENT_TYPE_VAR) {
+          // The method was defined with a method-level generic type argument from the caller. Return the TypeSpec token for the `M#` MVAR description, or
+          // The method was defined with a type-level generic type argument from the caller. Return the TypeSpec token for the `T#` VAR description
+          hr = metadata_emit->GetTokenFromTypeSpec(p_current_byte, 2,
+                                              ret_type_token);
+          return SUCCEEDED(hr);
+        } else if (*p_current_byte == ELEMENT_TYPE_GENERICINST) {
+          p_current_byte += 1;
+          return *p_current_byte == ELEMENT_TYPE_VALUETYPE;
+        } else {
+          return ElementTypeIsAlwaysValueType(CorElementType(*p_current_byte));
```

Comment:
Same as above. I need to set `ret_type_token` before potentially returning `true`.

Reply:
Commit 6b381fbb42efc8e9892015da508267fa584730a2 includes all the changes necessary to return the correct TypeRef for the value type.